### PR TITLE
upgrade perforce packages: p4, p4v

### PR DESCRIPTION
**Pull request by corngood**
###### Motivation for this change

This still doesn't have a permanent url for the specific minor version, and will break whenever they update. :(

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Download link is no longer available.

**Comment by corngood** [comment]
Yeah, looks like they mysteriously recalled 20.2:

https://www.perforce.com/support/software-release-index#tab-panel-6208
> Release 2020.2 *Temporarily Unavailable for Download

**Comment by SuperSandro2000** [comment]
What now? Update to a newer version? Update the URL? Can this PR be closed?

**Comment by corngood** [comment]
> Note: The 20.2 release of Helix Core Suite has been temporarily removed as we perform some Root Cause Analysis on customer reported issues.

Once they post a new version, I'll update this PR.

**Comment by corngood** [comment]
I switched both of these to archive.org, as suggested in #131292.

**Review by nioncode — APPROVED** [review]
Looks good to me!

File: pkgs/applications/version-management/p4v/default.nix [hunk 1]
```diff
@@ -1,12 +1,39 @@
...
+  qtdeclarative,
+  qtwebchannel,
+  syntax-highlighting,
+  openssl,
+  xkeyboard_config,
+  patchelfUnstable,
+  wrapQtAppsHook,
+  writeText
+}:
+let
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/p4v/default.nix:16]:
```suggestion
{ stdenv
, fetchurl
, lib,
, qtbase,
, qtwebengine
, qtdeclarative
, qtwebchannel
, syntax-highlighting
, openssl
, xkeyboard_config
, patchelfUnstable
, wrapQtAppsHook
, writeText
}:
let
```

**Comment by corngood** [comment]
Updated again, and applied your suggestions @SuperSandro2000.

File: pkgs/applications/version-management/p4v/default.nix [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
+  # expects it to succeed. The function will fail if CRYPTO_malloc has already
+  # been called, which happens at init time via qtwebengine -> ... -> libssh. I
+  # suspect it was meant to work with a version of Qt where openssl is
+  # statically linked or some other library is used.
+  crypto-hack = writeText "crypto-hack.c" ''
+      #include <stddef.h>
+      int CRYPTO_set_mem_functions(
+            void *(*m)(size_t, const char *, int),
+            void *(*r)(void *, size_t, const char *, int),
+            void (*f)(void *, const char *, int)) { return 1; }
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/p4v/default.nix:26]: I don't feel comfortable merging this with this patch. I don't know if it breaks crypto or not.

**Comment by nathyong** [comment]
Trying to test this now, but I don't seem to be able to download the sources from archive.org.

File: pkgs/applications/version-management/p4v/default.nix (hunk 2 above)
Comment (corngood) [reply pkgs/applications/version-management/p4v/default.nix:26]: We could try to build a modified Qt instead.  p4v is still a binary blob either way.

**Comment by corngood** [comment]
@nathyong What's the problem exactly?  I found archive.org extremely slow, and p4v is like 130m.  Any other ideas for mirrors are welcome.

File: pkgs/applications/version-management/p4v/default.nix (hunk 2 above)
Comment (corngood) [reply pkgs/applications/version-management/p4v/default.nix:26]: I tested latest with this hack disabled, and it still fails.  The error you get is:

```
Connect to server failed; check $P4PORT.
Programming error: CRYPTO_set_mem_functions(): Could not set OpenSSL allocation functions.
```

**Comment by nathyong** [comment]
I wasn't able to get any bytes from archive.org in the past. Seems fine (if a bit slow) now though.

**Comment by SuperSandro2000** [comment]
If the download link is not changing over time and is not gone again we shouldn't use web.archive.org.

**Comment by corngood** [comment]
The download link does change, unfortunately.  Perforce only provides a link to the minor version, which always points to the latest patch release.

Also, I had a good look at using the shipped qt libs in case it allowed the crypto hack to be removed, but things got hacky very quickly and I wasn't able (so far) to fix a crash in qt web engine.  It was feeling like it would be a worse outcome.

**Comment by Artturin** [comment]
@ofborg build p4v

**Comment by Artturin** [comment]
files that should not be in bin are in bin 
```
[nix-shell:~/.cache/nixpkgs-review/pr-110168]$ ls ./results/p4v/bin
p4admin  p4admin.bin  p4merge  p4merge.bin  p4v  p4v.bin  p4vc  qt.conf  QtWebEngineProcess
```

**Comment by corngood** [comment]
@Artturin yeah, that's not ideal, but it's unchanged in this upgrade and it's how they're distributed upstream.

**Comment by Artturin** [comment]
did not test

**Pull request merged**